contracts/auction/ClockAuction.sol

```diff
@@ -99,15 +141,26 @@ contract ClockAuction is AuctionRelated {
     // @dev bidder must use RING.transfer(address(this), _valueInRING, bytes32(_tokenId)
     // to invoke this function
     // @param _data - need to be generated from bytes32(tokenId)
-    function tokenFallback(address _from, uint256 _valueInRING, bytes _data) public whenNotPaused {
+
+    function tokenFallback(address _from, uint256 _valueInToken, bytes _data) public whenNotPaused {
+        uint256 tokenId = bytesToUint256(_data);
         if (msg.sender == address(RING)) {
             // assure it can be converted into uint256 correctly
             require(_data.length == 32);
-            uint256 tokenId = bytesToUint256(_data);
-
-            _bidWithRING(_from, tokenId, _valueInRING);
+            // TODO: add restriction
+            require(!landData.isReserved(tokenId));
+            _bidWithRING(_from, tokenId, _valueInToken);
         }
 
+        if (msg.sender == address(KTON)) {
+            // assure it can be converted into uint256 correctly
+            require(_data.length == 32);
+            // TODO: add restriction
+            // require tokenId is reserved
+            require(landData.isReserved(tokenId));
```

Comment:
Suggest that do not make strong connection between land and buy logics, this part can be moved to business logic rather than protocol level here.